File: plsql-checks/src/main/java/br/com/felipezorzo/sonar/plsql/checks/TooManyRowsHandlerCheck.java
/*
 * Sonar PL/SQL Plugin (Community)
 * Copyright (C) 2015 Felipe Zorzo
 * felipe.b.zorzo@gmail.com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
 */
package br.com.felipezorzo.sonar.plsql.checks;

import java.util.List;

import org.sonar.api.server.rule.RulesDefinition;
import org.sonar.check.Priority;
import org.sonar.check.Rule;
import org.sonar.squidbridge.annotations.ActivatedByDefault;
import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;

import com.sonar.sslr.api.AstNode;

import br.com.felipezorzo.sonar.plsql.api.PlSqlGrammar;

@Rule(
    key = TooManyRowsHandlerCheck.CHECK_KEY,
    priority = Priority.CRITICAL
)
@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.ERRORS)
@SqaleConstantRemediation("5min")
@ActivatedByDefault
public class TooManyRowsHandlerCheck extends AbstractBaseCheck {
    public static final String CHECK_KEY = "TooManyRowsHandler";

    @Override
    public void init() {
        subscribeTo(PlSqlGrammar.EXCEPTION_HANDLER);
    }

    @Override
    public void visitNode(AstNode node) {
        // is a TOO_MANY_ROWS handler
        AstNode exceptionName = node.getChildren().get(1);
        
        if (exceptionName.is(PlSqlGrammar.PRIMARY_EXPRESSION)) {
            exceptionName = exceptionName.getFirstChild();
        }
        
        if (exceptionName.is(PlSqlGrammar.IDENTIFIER_NAME) && 
                "TOO_MANY_ROWS".equalsIgnoreCase(exceptionName.getTokenValue())) {
            // and have only one NULL_STATEMENT
            List<AstNode> children = node.getChildren(PlSqlGrammar.STATEMENT);
            if (children.size() == 1 && children.get(0).getFirstChild().is(PlSqlGrammar.NULL_STATEMENT)) {
                getContext().createLineViolation(this, getLocalizedMessage(CHECK_KEY), node);
            }
            
        }
    }
}

File: zpa-checks/src/main/kotlin/com/felipebz/zpa/checks/TooManyRowsHandlerCheck.kt
/**
 * Z PL/SQL Analyzer
 * Copyright (C) 2015-2025 Felipe Zorzo
 * mailto:felipe AT felipezorzo DOT com DOT br
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.felipebz.zpa.checks

import com.felipebz.flr.api.AstNode
import com.felipebz.zpa.isOf
import com.felipebz.zpa.sslr.NullStatement
import com.felipebz.zpa.typeIs
import com.felipebz.zpa.api.PlSqlGrammar
import com.felipebz.zpa.api.annotations.*

@Rule(priority = Priority.CRITICAL, tags = [Tags.BUG])
@ConstantRemediation("5min")
@RuleInfo(scope = RuleInfo.Scope.ALL)
@ActivatedByDefault
class TooManyRowsHandlerCheck : AbstractBaseCheck() {

    override fun init() {
        subscribeTo(PlSqlGrammar.EXCEPTION_HANDLER)
    }

    override fun visitNode(node: AstNode) {
        // is a TOO_MANY_ROWS handler
        val exceptions = node.getChildren(PlSqlGrammar.VARIABLE_NAME)

        for (exception in exceptions) {
            val child = exception.firstChild

            if (child.typeIs(PlSqlGrammar.IDENTIFIER_NAME) && "TOO_MANY_ROWS".equals(child.tokenValue, ignoreCase = true)) {
                // and have only one NULL_STATEMENT
                val children = node.getFirstChild(PlSqlGrammar.STATEMENTS).children
                if (children.size == 1 && children[0].isOf<NullStatement>()) {
                    addIssue(node, getLocalizedMessage())
                }
            }
        }
    }

}

